### PR TITLE
Feature：role

### PR DESCRIPTION
**Pull request by 15168440402**
### Features:
-- Add Role  [#19](https://github.com/masastack/MASA.Auth/pull/19)
-- Add RBAC3  [#19](https://github.com/masastack/MASA.Auth/pull/19)

### CODE REVIEW

File: src/Contracts/Masa.Auth.Contracts.Admin/Permissions/RoleSelectDto.cs [hunk 1]
```diff
@@ -6,9 +6,12 @@ public class RoleSelectDto
 
     public string Name { get; set; }
 
-    public RoleSelectDto(Guid id, string name)
+    public int QuantityAvailable { get; set; }
```
Comment (doddgu) [inline src/Contracts/Masa.Auth.Contracts.Admin/Permissions/RoleSelectDto.cs:9]: QuantityAvailable -> AvailableQuantity

File: src/Services/Masa.Auth.Service.Admin/Domain/Permissions/Aggregates/Role.cs [hunk 2]
```diff
@@ -10,38 +19,77 @@ public class Role : AuditAggregateRoot<Guid, Guid>, ISoftDelete
...
+        set
+        {
+            if (value < 0)
+                throw new UserFriendlyException("This operation cannot be completed due to role Limit restrictions");
+
+            _limit = value;
+        }
+    }
+
+    public int QuantityAvailable { get; private set; }
```
Comment (doddgu) [inline src/Services/Masa.Auth.Service.Admin/Domain/Permissions/Aggregates/Role.cs:34]: Saame as before

File: src/Services/Masa.Auth.Service.Admin/Domain/Permissions/EventHandler/UpdateRoleLimitDomainEventHandler.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            _authDbContext = authDbContext;
+        }
+
+        [EventHandler(1)]
+        public async Task UpdateRoleLimitAsync(UpdateRoleLimitDomainEvent roleEvent)
+        {
+            var roles = await _authDbContext.Set<Role>()
+                                    .Include(r => r.Users)
+                                    .Include(r => r.Teams)
+                                    .ThenInclude(tr => tr.Team)
```
Comment (doddgu) [inline src/Services/Masa.Auth.Service.Admin/Domain/Permissions/EventHandler/UpdateRoleLimitDomainEventHandler.cs:22]: tr?

File: src/Services/Masa.Auth.Service.Admin/Domain/Permissions/EventHandler/UpdateRoleLimitDomainEventHandler.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                                    .ThenInclude(t => t.TeamStaffs)
+                                    .Where(r => roleEvent.Roles.Contains(r.Id))
+                                    .ToListAsync();
+            foreach (var role in roles)
+            {
+                var quantityAvailable = GetQuantityAvailable(role);
+                if (role.Limit == 0 || quantityAvailable >= 0)
```
Comment (doddgu) [inline src/Services/Masa.Auth.Service.Admin/Domain/Permissions/EventHandler/UpdateRoleLimitDomainEventHandler.cs:29]: quantityAvailable = 0?

File: src/Services/Masa.Auth.Service.Admin/Domain/Permissions/EventHandler/UpdateRoleLimitDomainEventHandler.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            await _roleRepository.UpdateRangeAsync(roles);
+
+            int GetQuantityAvailable(Role role)
+            {
+                var quantityAvailable = role.Limit - role.Users.Count;
+                if (role.Teams.Count > 0)
+                {
+                    foreach (var teamRole in role.Teams)
+                    {
+                        if (teamRole.TeamMemberType == TeamMemberTypes.Admin)
```
Comment (doddgu) [inline src/Services/Masa.Auth.Service.Admin/Domain/Permissions/EventHandler/UpdateRoleLimitDomainEventHandler.cs:44]: ????

File: src/Web/Masa.Auth.Web.Admin.Rcl/wwwroot/i18n/zh-CN.json [hunk 6]
```diff
@@ -133,13 +134,24 @@
...
+  "RoleOwner": "角色使用者",
+  "Limit the number of bindings": "绑定限制次数",
+  "The number of bindings cannot be less than ": "绑定限制次数不能小于",
   "Failed to query roleList data:": "查询角色列表数据失败:",
   "Failed to query roleDetail data:": "查询角色详情数据失败:",
   "Add role data success": "添加角色数据成功",
   "Failed to add role:": "添加角色数据失败:",
   "Edit role data success": "编辑角色数据成功",
   "Failed to edit role": "编辑角色数据失败:",
+  "This operation cannot be completed due to role Limit restrictions": "由于角色Limit限制，无法完成此操作",
```
Comment (doddgu) [inline src/Web/Masa.Auth.Web.Admin.Rcl/wwwroot/i18n/zh-CN.json:152]: * Limit restrictions -> cardinality constraint.
* Limit限制 -> 基数约束

File: src/Services/Masa.Auth.Service.Admin/Domain/Permissions/EventHandler/UpdateRoleLimitDomainEventHandler.cs (hunk 4 above)
Comment (15168440402) [reply src/Services/Masa.Auth.Service.Admin/Domain/Permissions/EventHandler/UpdateRoleLimitDomainEventHandler.cs:29]: quantityAvailable = 0 means the character cardinality constraint has been used up

**Review by doddgu — APPROVED** [review]
LGTM

**Pull request merged**